FLIGHT_DEVELOPER_GUIDE.md

```diff
@@ -0,0 +1,442 @@
+
+#### Pitfall 1: deleting someone else's object
+Imagine that you want to create a thing named "foo". The thing might be a file, a cloud
+object, a database row, whatever. You write a step like this:
+
+```
+Step: Create a thing
+ do():
+   create "foo"
+ undo():
+   delete "foo"
+```
+But what if "foo" already exists when this step is run? The execution will go:
+ 1. Stairway runs `do()` and call create "foo". That results in a "foo already exists"
+ exception.
+ 1. Stairway catches the exception and decides the flight is failed. It
+ transitions to undoing.
+ 1. Stairway runs `undo()` and calls delete "foo". It deletes the pre-existing object.
+
+#### Pitfall 2: failing when you already succeeded
+
+```
+Step: Create a thing - take 2
+ do():
+   if "foo" exists {
+     write "foo already existed" to the working map
+     throw "foo exists"
+   }
+   create "foo"
+ undo():
+   if the working map does says "foo already existed" {
+     then return success
+   }
+   delete "foo"
+```
+That is better. When Stairway catches the exception and transitions to undoing, it writes
+the flight state, so the working map gets written. The undo can skip the delete. Yay!
+
+The problem is that the `do()` is no longer idempotent. Suppose there is a server failure right
+after the "foo" thing is created, but before Stairway records the step result. On
+recovery, Stairway will rerun the `do()` step. It will say, "foo already existed", and the
+flight will fail.
+
+#### Pitfall 3: concurrent operation
+
+A more reliable way to accomplish the existence check is to do it in a separate step. So
+now you have this pseudo-code:
+```
+Step 1: Check thing existence
+ do():
+   if "foo" exists {
+     error out with "foo already exists"
+   }
+ undo():
+   no action
+
+Step 2: Create thing - we only get to this step if "foo" does not exist
+ do():
+   create "foo" - ignore "foo" already exists error
+ undo():
+   delete "foo"
+```
+This works pretty well, but there is still a problem. Suppose there are two flights
+running in parallel that are both trying to create "foo". They will both make it through
+Step 1 and decide "foo" doesn't exist. Then they will both try to create "foo". One will
+successfully create _their version_ of "foo". The other will get a "foo already
+exists" error, ignore it, and think they have succeeded.
+
+#### A Working Approach
+
+The typical case in our system is that we are creating a cloud resource and recording that
+resource in a database. We can use the database state to address the pitfalls. The general
+approach is:
+ 1. Create or provide as flight input a random UUID
+ 1. Write a row in the database with the UUID and the name. If there is a name
+    collision, we fail in this step and do not try to create anything. If there is a UUID
+    collision, we know that it was this flight that previously wrote the row and we can proceed to the next step.
+    If all concurrent creators check the database first, they will conflict on the name
```

Comment:
how do they conflict on the name before they've written anything to the db?

Reply:
The conflict on the UIUID and name in the process of writing to the db.